### ISSUE TITLE
Need help understanding example Code

### ISSUE THREAD

**Issue by atburghardt** [Novalia/NovaliaBLE#2]
First of all, this Code is part of a Cordova-Plugin for the Novalia-framework.
After starting initinterfaceIfNeeded the App starts looking for Devices. When I start my Testboard, the App crashes in onDeviceDiscovered when trying to take an Object of the empty selectedDevices-array . 

XCode updated some parts, because of deprecated Functions. Can you please help me understand, how to connect to you Boards correctly? Is there some more Documentation?
My Code:
`
    import NovaliaBLE

    @objc(NovaliaSwift) class NovaliaSwift : CDVPlugin , NovaliaBLEInterfaceDelegate, NovaliaBLEDeviceEventDelegate {

      var interface: NovaliaBLEInterface!
      var previousBLEState: NovaliaBLEState!
      var launchDiscoveryTimer: Timer!
      var selectedDevices = [NovaliaBLEDevice]()
      var currentDevice: NovaliaBLEDevice!

 

    // MARK: NovaliaBLEInterface methods
     @objc(initInterfaceIfNeeded:)
      func initInterfaceIfNeeded(command: CDVInvokedUrlCommand) {
          print("initInterfaceIfNeeded")
          if(interface == nil) {
              print("no Interface")
              interface = NovaliaBLEInterface(delegate: self)
              previousBLEState = BLEStateNotReady
              launchDiscoveryTimer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(NovaliaSwift.startDiscovery), userInfo: nil, repeats: false)
          }
      }

      func startDiscovery() {
          print("startDiscovery")
          launchDiscoveryTimer = Timer.scheduledTimer(timeInterval: 30.0, target: self, selector: #selector(NovaliaSwift.stopDiscovery), userInfo: nil, repeats: false)

          if currentDevice == nil {
            print("No device connected")
              // Set a status message to indicate no device connected
          }

          // Set the device name to be searched for e.g. novppia1
          // Setting the device name to * searches for all devices
          print("look for device novppia1")
          interface.startDeviceDiscovery("novppia1")
      }

      func stopDiscovery() {
          print("stopDiscovery")
          if currentDevice == nil {
            print("Set a status message to indicate no device connected")
              // Set a status message to indicate no device connected
          }

          launchDiscoveryTimer = nil
          interface.stopDeviceDiscovery()
      }

    // MARK: NovaliaBLEInterface Delegate methods
    func onDeviceDiscovered(_ device: NovaliaBLEDevice!) {

        print("Device discovered")
        if currentDevice == nil {
            // Set a status message to indicate connecting to device
        }

        var array = selectedDevices
        let index = array.index(of: device)

        if(index != nil) {
            array[index!] = device
        } else {
            array.append(device)
        }

        device.delegate = self
        selectedDevices = array
        interface.connect(to: device)
    }

    func onDeviceListChanged(newList: [AnyObject]!) {
        print("onDeviceListChanged")
        var connected = 0
        var connecting = 0
        var disconnected = 0

        for device: NovaliaBLEDevice in newList as! [NovaliaBLEDevice] {
            if(isDeviceConnected(device: device)) {
                print("(device.uuid.UUIDString) connected")
                connected += 1
            } else if(isDeviceConnecting(device: device)) {
                print("(device.uuid.UUIDString) connecting")
                connecting += 1
            } else if(isDeviceDisconnected(device: device)) {
                print("(device.uuid.UUIDString) disconnected")
                disconnected += 1
            }
        }

        // This is where
      //  an indicator of how many devices are connected, connecting, disconnected should be displayed
        print(String(format:"Connected %d Connecting %d Disconnected %d", connected, connecting, disconnected))

        let isConnecting = (connecting > 0)

        if(isConnecting) {
            // Start any connecting animation
        } else {
            // Stop any connecting animation
        }
    }

    func onDeviceConnected(_ device: NovaliaBLEDevice!) {
        // Set a status message to indicate device connected
        print("CONNECTED TO DEVICE")
        if(device.deviceName == "novppia1") {
            currentDevice = device
            if(isDeviceConnected(device: currentDevice)) {
                print("(currentDevice.uuid.UUIDString) connected")
            } else if(isDeviceConnecting(device: currentDevice)) {
                print("(currentDevice.uuid.UUIDString) connecting")
            } else if(isDeviceDisconnected(device: currentDevice)) {
                print("(currentDevice.uuid.UUIDString) disconnected")
            }
        } else {
            print("\(device.deviceName) is not what we are looking for")
        }
    }

    func isDeviceConnected(device: NovaliaBLEDevice) -> Bool {
        print("isDeviceConnected")
        return (device.status & NovaliaBLEDeviceConnected) == NovaliaBLEDeviceConnected
    }
    func isDeviceConnecting(device: NovaliaBLEDevice) -> Bool {
        print("isDeviceConnecteing")

        return (device.status & NovaliaBLEDeviceConnecting) == NovaliaBLEDeviceConnecting
    }
    private func isDeviceDisconnected(device: NovaliaBLEDevice) -> Bool {
        print("Private isDeviceConnected")

        return isDeviceConnected(device: device) == false && isDeviceConnecting(device: device) == false
    }

    func onDeviceDisconnected(_ device: NovaliaBLEDevice!) {
        print("onDeviceDIsconnected")
        currentDevice = nil

        // Set a status message to indicate no device connected

        // Try to reconnect
        interface.connect(to: device)
    }

    func onBLEStateChanged(_ state: NovaliaBLEState) {
        print("onBleStateCHanged")
        if(previousBLEState == state) {
            return
        }

        previousBLEState = state

        var title = ""
        var message = ""

        switch state.rawValue {
        case BLEStateOff.rawValue:
            title = "Bluettoth Power"
            message = "You must turn on Bluetooth in Settings in order to use this application"
        case BLEStateUnsupported.rawValue:
            title = "Bluetooth Unsupported"
            message = "Your device does not seem to be compatible with Bluetooth Low Energy standard."
        case BLEStateNotReady.rawValue:
            title = "Bluetooth Not Ready"
            message = "Bluetooth service on your device does not seem to be ready. It may be initialising or restarting."
        default: ()
        }

        if(!title.isEmpty) {
            let alert = UIAlertController(title: title, message: message, preferredStyle: UIAlertControllerStyle.alert)
            alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil))
            self.viewController.present(alert, animated: true, completion: nil)
        }
    }

    func onButtonPressed(_ button: Int32, velocity: Int32, onDevice device: Any!) {
        print("BUTTON PRESSED")
        print("Button pressed \(button) on device \((device as AnyObject).uuidString)")

        // Now do something in response to a button being pressed
    }
}`

**Comment by andrewsage** [comment]
The repo has now been updated to the latest version. Can you please let me know if it still crashes?

**Comment by atburghardt** [comment]
I updated and copied the readme code again. XCode wants me to change some Functions again, because they have been renamed:

NSTimer -> Timer
scheduledTimerWithTimeInterval(  -> scheduledTimer(timeInterval:
array.indexOf(device)    ->   array.index(of: device)
 interface.connectToDevice(device)   ->     interface.connect(to: device)
if(isDeviceConnected(device))    ->      if(isDeviceConnected(device: device)) {
func onDeviceDiscovered(device: NovaliaBLEDevice!) {      ->   func onDeviceDiscovered(_ device: NovaliaBLEDevice!) {

This throws an exeption: 
 print("\(device.uuid.UUIDString) connected")   ->   Value of type UUID has no member UUIDString
deleted this part, for the time beeing


Still in the function onDeviceDiscovered, it crashes at 
        let index = array.index(of: device)

 

     func onDeviceDiscovered(_ device: NovaliaBLEDevice!) {
        
        print("Device discovered")
        if currentDevice == nil {
            // Set a status message to indicate connecting to device
        }
        
        var array = selectedDevices
        let index = array.index(of: device)
        
        if(index != nil) {
            array[index!] = device
        } else {
            array.append(device)
        }
        
        device.delegate = self
        selectedDevices = array
        interface.connect(to: device)
    }

**Comment by andrewsage** [comment]
Can you get the example app included with the CocoPod to work?
The code in it may be more up to date than the read me.

**Comment by atburghardt** [comment]
Hey, the example app is working now.  I copied the Code to my plugin, code is up to date. Thanks!

I want to implement a Cordova-Plugin: there is no UIViewController in my NovaliaSwift-Plugin-Code -> I deleted all Interface Interaction (Label and Button updates).
I run my app and start the function ifInterfaceNeeded manually by terminal. When discovering the Tablet, it still crashes in onDeviceDiscovered with : Thread 1: EXC_BAD_ACCESS(code=1, address=0x10)

        var array = selectedDevices        
        let index = array.index(of: device)

I don't get why it crashes there. If it cant find the device in the array, it should return nil and keep working. Any Idea what causes the crash?


Edit:
my Class looks like this:

class NovaliaSwift : CDVPlugin, NovaliaBLEInterfaceDelegate, NovaliaBLEDeviceEventDelegate {
